src/utils.rs

```diff
@@ -10,177 +23,154 @@ pub fn has_attribute(needle: &str, attributes: &[syn::Attribute]) -> bool {
-                if group.delimiter() != proc_macro2::Delimiter::Parenthesis {
-                    continue;
-                }
-
-                let name_value_pairs = ::syn::parse::Parser::parse2(
-                    Punctuated::<MetaNameValue, Token![,]>::parse_terminated,
-                    group.stream(),
-                );
-
-                if name_value_pairs.is_err() {
-                    let comma_seperated_values = ::syn::parse::Parser::parse2(
-                        Punctuated::<syn::Expr, Token![,]>::parse_terminated,
-                        group.stream(),
-                    );
-
-                    if comma_seperated_values.is_err() {
-                        continue;
-                    }
-
-                    let comma_seperated_values = comma_seperated_values.unwrap();
-
-                    for comma_seperated_value in comma_seperated_values {
-                        match comma_seperated_value {
-                            syn::Expr::Path(expr_path) => {
-                                let segments = expr_path.path.segments;
-
-                                if segments.is_empty() {
-                                    continue;
-                                }
-
-                                if segments[0].ident.to_string().eq(arg) {
-                                    found = true;
-                                    value = String::from(arg);
-
-                                    break;
-                                }
-                            }
-                            _ => continue,
-                        }
-                    }
-
-                    continue;
-                }
-
-                let name_value_pairs = name_value_pairs.unwrap();
-
-                for name_value_pair in name_value_pairs {
-                    if name_value_pair.path.is_ident(arg) {
-                        found = true;
-                        value = name_value_pair.value.to_token_stream().to_string();
-                        // removes quotes around the value
-                        value = value[1..value.len() - 1].to_string();
-
-                        break;
-                    }
-                }
-            }
-        }
-
-        if found {
-            return Some(value);
-        } else {
-            return None;
-        }
-    }
-    None
+    // check if attribute list contains the argument we are interested in
+    // TODO: don't use a for loop here or iterator here
+    get_attribute(needle, attributes).and_then(|attr| {
+        attr.meta
+            .to_token_stream()
+            .into_iter()
+            .filter_map(|token| check_token(token, arg))
+            .next()
+    })
 }
 
 /// Check has an attribute arg.
 pub fn has_attribute_arg(needle: &str, arg: &str, attributes: &[syn::Attribute]) -> bool {
     get_attribute_arg(needle, arg, attributes).is_some()
 }
 
+fn check_token_tree(tt: proc_macro2::TokenTree) -> Option<String> {
```

Comment:
could we also add a comment on how and when to use this method?

Similar to:

```
// this detects the '(...)' part in #[serde(rename_all = "UPPERCASE", tag = "type")]
// we can use this to get the value of a particular argument
// or to see if it exists at all
```

Reply:
Sure, no problem. I took a shot in the latest commits. Please let me know what you think.